pkg/storage/export/export/vm-source.go

```diff
@@ -0,0 +1,244 @@
+		log.Log.V(3).Infof("Processing VM %s", key)
+		keys, err := ctrl.VMExportInformer.GetIndexer().IndexKeys("vm", key)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+		for _, key := range keys {
+			ctrl.vmExportQueue.Add(key)
+		}
+	}
+}
+
+func (ctrl *VMExportController) handleVMI(obj interface{}) {
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+
+	if vmi, ok := obj.(*virtv1.VirtualMachineInstance); ok {
+		vm := ctrl.getVMFromVMI(vmi)
+		if vm != nil {
+			key, _ := cache.MetaNamespaceKeyFunc(vm)
+			log.Log.V(3).Infof("Processing VM %s", key)
+			keys, err := ctrl.VMExportInformer.GetIndexer().IndexKeys("vm", key)
+			if err != nil {
+				utilruntime.HandleError(err)
+				return
+			}
+			for _, key := range keys {
+				ctrl.vmExportQueue.Add(key)
+			}
+			return
+		}
+		pvcs := ctrl.getPVCsFromVMI(vmi)
+		for _, pvc := range pvcs {
+			ctrl.handlePVC(pvc)
+		}
+	}
+}
+
+func (ctrl *VMExportController) getPVCsFromVMI(vmi *virtv1.VirtualMachineInstance) []*corev1.PersistentVolumeClaim {
+	var pvcs []*corev1.PersistentVolumeClaim
+	for _, volume := range vmi.Spec.Volumes {
+		pvcName := storagetypes.PVCNameFromVirtVolume(&volume)
+		if pvc := ctrl.getPVCsFromName(vmi.Namespace, pvcName); pvc != nil {
+			pvcs = append(pvcs, pvc)
+		}
+	}
+	return pvcs
+}
+
+func (ctrl *VMExportController) getOwnerVMexportKey(obj metav1.Object) string {
+	ownerRef := metav1.GetControllerOf(obj)
+	var key string
+	if ownerRef != nil {
+		if ownerRef.Kind == exportGVK.Kind && ownerRef.APIVersion == exportGVK.GroupVersion().String() {
+			key = controller.NamespacedKey(obj.GetNamespace(), ownerRef.Name)
+		}
+	}
+	return key
+}
+
+func (ctrl *VMExportController) getVMFromVMI(vmi *virtv1.VirtualMachineInstance) *virtv1.VirtualMachine {
+	ownerRef := metav1.GetControllerOf(vmi)
+	if ownerRef != nil {
+		if ownerRef.Kind == "VirtualMachine" && ownerRef.APIVersion == virtv1.GroupVersion.String() {
+			if vm, exists, err := ctrl.getVm(vmi.Namespace, ownerRef.Name); !exists || err != nil {
+				log.Log.V(3).Infof("Unable to get owner VM %s/%s for VMI %s/%s", vmi.Namespace, ownerRef.Name, vmi.Namespace, vmi.Name)
+			} else {
+				return vm
+			}
+		}
+	}
+	return nil
+}
+
+func (ctrl *VMExportController) isSourceAvailableVM(vmExport *exportv1.VirtualMachineExport) (bool, string, error) {
+	vmi, exists, err := ctrl.getVmi(vmExport.Namespace, vmExport.Spec.Source.Name)
+	if err != nil {
+		return false, "", err
+	}
+	if exists {
+		return false, fmt.Sprintf("Virtual Machine %s/%s is running", vmi.Namespace, vmi.Name), nil
+	}
+	return true, "", nil
+}
+
+func (ctrl *VMExportController) getPVCFromSourceVM(vmExport *exportv1.VirtualMachineExport) (*sourceVolumes, error) {
+	pvcs, allPopulated, err := ctrl.getPVCsFromVM(vmExport.Namespace, vmExport.Spec.Source.Name)
+	if err != nil {
+		return &sourceVolumes{}, err
+	}
+	if pvcs != nil && !allPopulated {
```

Comment:
awkward to check if a slice is != nil.  Checking length is the idiomatic way

Reply:
Done